File: src/ts/test/corpus.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { DECLARATIONS, decodeDeclaration } from "../dist/corpus.js";
import { corpus } from "./helpers.mjs";

test("every corpus instance gets the declared verdict from the TypeScript decoder", async () => {
  const all = await corpus();
  assert.ok(all.length >= 70);
  const failures = [];
  for (const { declaration, expectation, file, json } of all) {
    assert.ok(DECLARATIONS.includes(declaration), `unknown declaration dir ${declaration}`);
    let verdict = "accepted";
    try { decodeDeclaration(declaration, json); } catch { verdict = "rejected"; }
    if (verdict !== expectation) failures.push(`${declaration}/${expectation}/${file}: typescript said ${verdict}`);
  }
  assert.deepEqual(failures, []);
});

test("every declaration has valid and invalid coverage", async () => {
  const all = await corpus();
  for (const declaration of DECLARATIONS) {
    assert.ok(all.some((c) => c.declaration === declaration && c.expectation === "accepted"), `${declaration} valid`);
    assert.ok(all.some((c) => c.declaration === declaration && c.expectation === "rejected"), `${declaration} invalid`);
  }
});

test("qualified declaration names are accepted", () => {
  assert.equal(decodeDeclaration("OresDnd.DndOperation", '"copy"'), "copy");
  assert.throws(() => decodeDeclaration("OresDnd.Nope", "{}"), /unknown declaration/);
});

File: src/dart/test/corpus_test.dart
import 'package:ores_dnd/ores_dnd.dart';
import 'package:test/test.dart';

import 'helpers.dart';

void main() {
  final all = corpus();

  test(
    'every corpus instance gets the declared verdict from the Dart decoder',
    () {
      expect(all.length, greaterThanOrEqualTo(70));
      final failures = <String>[];
      for (final c in all) {
        expect(
          declarations,
          contains(c.declaration),
          reason: 'unknown declaration dir ${c.declaration}',
        );
        var verdict = 'accepted';
        try {
          decodeDeclaration(c.declaration, c.json);
        } on FormatException {
          verdict = 'rejected';
        }
        if (verdict != c.expectation)
          failures.add(
            '${c.declaration}/${c.expectation}/${c.file}: dart said $verdict',
          );
      }
      expect(failures, isEmpty);
    },
  );

  test('every declaration has valid and invalid coverage', () {
    for (final declaration in declarations) {
      expect(
        all.any(
          (c) => c.declaration == declaration && c.expectation == 'accepted',
        ),
        isTrue,
        reason: '$declaration valid',
      );
      expect(
        all.any(
          (c) => c.declaration == declaration && c.expectation == 'rejected',
        ),
        isTrue,
        reason: '$declaration invalid',
      );
    }
  });

  test('qualified declaration names are accepted', () {
    expect(
      decodeDeclaration('OresDnd.DndOperation', '"copy"'),
      DndOperation.copy,
    );
    expect(
      () => decodeDeclaration('OresDnd.Nope', '{}'),
      throwsFormatException,
    );
  });
}
